src/Catty/ViewController/Upload/UploadViewController.swift

```diff
@@ -259,171 +203,37 @@ class UploadViewController: UIViewController {
-                        debugPrint("Error: ", serverResponse ?? "undefined")
-                        DispatchQueue.main.async(execute: {
-                            if serverResponse != nil {
-                                Util.alert(withText: serverResponse)
-                            } else {
-                                Util.alert(withText: kLocalizedUploadProblem)
-                            }
-                        })
-
-                        if statusCode == self.statusCodeTokenWrong {
-                            //Token not valid
-                            UserDefaults.standard.set(false, forKey: kUserIsLoggedIn)
-
-                            var viewArray: [AnyHashable]?
-                            if let aControllers = self.parent?.navigationController?.viewControllers {
-                                viewArray = aControllers
-                            }
-                            viewArray?.removeLast()
-                            var newViewArray: [AnyHashable]?
-                            if let anArray = viewArray {
-                                newViewArray = anArray
-                            }
-                            if let anArray = newViewArray as? [UIViewController] {
-                                self.parent?.navigationController?.setViewControllers(anArray, animated: true)
-                            }
-                        }
-                    }
-                }
-            }
-            dataTask.resume()
-        } else {
-            debugPrint("Could not build checksum")
-            enableUploadView()
-            Util.alert(withText: kLocalizedUploadProblem)
+        if let project = project, let uploader = self.uploader {
+            uploader.upload(project: project,
+                            completion: { error in
+                                self.enableUploadView()
+                                if let error = error {
+                                    switch error {
+                                    case .unexpectedError, .timeout:
+                                        Util.defaultAlertForNetworkError()
+                                    case .zippingError, .invalidProject, .request:
+                                        Util.alert(withText: kLocalizedUploadProblem)
+                                    case .authenticationFailed:
+                                        UserDefaults.standard.set(false, forKey: kUserIsLoggedIn)
+
+                                        let alert = UIAlertController(title: kLocalizedPocketCode, message: kLocalizedSessionExpired, preferredStyle: .alert)
+                                        alert.view.tintColor = UIColor.globalTint
+                                        alert.view.backgroundColor = UIColor.clear
+
+                                        let alertAction = UIAlertAction(title: kLocalizedOK, style: .cancel) { _ in
```

Comment:
Please use our `AlertControllerBuilder` instead of using the `UIAlertAction`, e.g.:

```
AlertControllerBuilder.alert(title: title, message: message)
            .addDefaultAction(title: buttonTitle) { completionHandler... }.build()
            .showWithController(self)
```